libr/util/buf.c

```diff
@@ -464,30 +483,38 @@ static int r_buf_fcpy_at (RBuffer *b, ut64 addr, ut8 *buf, const char *fmt, int
 		   tsize and m are not user controled, then don't
 		   need to check possible overflow.
 		 */
-		if (!UT64_ADD (&check_len, len, tsize*m))
-			return -1;
-		if (!UT64_ADD (&check_len, check_len, addr))
-			return -1;
+		if (!UT64_ADD (&check_len, new_len, tsize*m)) {
+			return -1;}
+		if (!UT64_ADD (&check_len, check_len, new_addr)) {
+			return -1;}
 		if (check_len > b->length) {
-			return check_len;
-			// return -1;
+			if (check_len < file_sz) {
+				iob->read_at (iob->io, b->base + new_addr + new_len, b->buf, b->length);
+				b->base += new_addr + new_len;//TODO: check for correctness
+				new_addr = 0;////TODO: take a look at this if it is set correctly.
+				b->cur = new_addr;
+				new_len = 0;
+			} else {
+				return check_len;
+			}
 		}
 
 		for (k = 0; k < m; k++) {
 			if (write) {
 				r_mem_swaporcopy ((ut8*)&buf[addr+len+(k*tsize)],
 						  (const ut8*)&b->buf[len+(k*tsize)],
-						  tsize, bigendian);
+						  tsize, bigendian);//XXX: This might mess up. Take a look at this.
```

Comment:
The correct way seems this: `r_mem_swaporcopy ((ut8*)&buf[addr+len+(k*tsize)],(const ut8*)&b->buf[new_len+(k*tsize)],tsize, bigendian);` because when b->buf will be replaced, new_len will be updated and it will start reading from the start of buf again. Just wanna confirm it though.